### ISSUE TITLE
should brigand go all eager

### ISSUE THREAD

**Issue by odinthenerd** [edouarda/brigand#209]
This seemed extreme to me when @brunocodutra made this change to metal but I am getting to understand it more and more. fast tracking `bind<eager_thing,_1>` is going to be faster than with lazy stuff and making things lazy in a generic way is possible but not the other way around. We should keep the lazy lambdas but the question is if every algorithm should also be available in a lazy form. I'm just floating the question, not sure I have made up my mind yet.

**Comment by edouarda** [comment]
Let's release 1.3 first :-)

**Comment by edouarda** [comment]
Did you reach a conclusion?

**Comment by odinthenerd** [comment]
Let's put it this way, the only things that speaks for lazy evaluation are:
- people being used to it
- complex lambdas are terser
- in cases like std::conditional both outcomes need not be evaluated

If we hypothesize that most TMP code has yet to be written the first argument is not so important, at the end of the day people are not writing a whole lot of complex lambdas even though they are cool and one can always lazy wrap an eager call.

On the other side (all eager)
- we have a purer library which is easier to teach if teaching from scratch
- instantiating types is more expensive than resolving aliases
- if you care about speed you are probably composing your lambdas manually, if you don't want speed your probably using hana

Essentially the only time one needs to create a type are:
- unpacking a list
- making a decision
- returning/trampolining (to flatten deep recursions)

We already have a generic unpacker call `wrap`, making decisions 'generic' is essentially the crux of my meta monad pattern, trampolining is not needed as often with good fast tracking and composing or chaining actions means we don't need to return intermediate results as often. 

In other words I think a truly all eager library can run considerably faster. On the other hand "truly all eager" would also mean many breaking changes and a break in compatibility to boost.mpl style.

**Comment by edouarda** [comment]
The thing is that it's very easy to switch from Boost.MPL to Brigand and I think it's a powerful feature.

**Comment by odinthenerd** [comment]
I agree! People need portability more than they need speed in the majority of cases. 

With current knowledge making the "fastest" possible library would end up looking little like brigand. I think the next step for me is to make a truly all eager experimental lib totally green field and see how far down the rabbit hole we can go and then with that knowledge rewrite the backend of brigand where possible.

**Comment by edouarda** [comment]
Indeed, if the speed gain is limited and/or visible for very large lists, then it means it's not that useful. It would also be interesting to see what C++ 17 implies in terms of meta programming.
